forem/forem
# Re-add mention autocomplete to comments, fixing bugs

## What type of PR is this? (check all applicable)


## Description

**This PR reverts the revert of mention autocomplete in comments, adding some bug fixes.**

After shipping #13061 we realised that a side effect of turning the whole text area into a Reach UI combobox was that navigation within the text area using the up/down arrow keys was no longer functioning. Further investigation showed that Reach UI calls `preventDefault` on these keypresses which is why they weren't registering.

This PR switches up our approach to get around this. The MentionAutocomplete component now creates **two** textareas:

- one vanilla textarea
- one combobox textarea which is initially hidden with `display 'none'`

When a user activates the autosuggest, the vanilla textarea is hidden, and the combobox takes over. When search is exited (either by selection, spacing away, or deleting all text including the `@` sign), the reverse happens. As a text area takes over, focus is shifted so that the user can keep typing without interruption.

This PR also adds a fix for an issue where editing a comment was not showing the original comment pre-populated in the comment box.

## Related Tickets & Documents

https://github.com/forem/rfcs/pull/16
#13061

## QA Instructions, Screenshots, Recordings

The changes here affect all the comment areas mentioned in the original PR:

- comments on posts
- replies to comments on posts
- replies to comments in notifications
- editing of those comments

Steps to test post comments:

- Go to a post
- Click to add a comment
- Start typing, including some @mentions (you might want to add some users with similar usernames to your local DB to be able to see fuller lists appearing in the dropdown)
- Experiment with selecting options with keyboard (up/down arrows + Enter) and mouse
- Experiment with deleting portions of the mention, ignoring the options and keeping typing, etc - try to break it 😄
- Repeat as above with replies to comments on a post

Steps to test post comment replies in notifications:

- Send a notification to your current user in the console:
```
user = User.find_by(username: "admin_mcadmin")
other_user = User.first
article = user.articles.last
comment = article.comments.create!(user: other_user, body_markdown: "Test something")
Notification.send_new_comment_notifications_without_delay(comment)
```
- Refresh the page and go to the notification
- Click reply and repeat the above steps

In all cases, experiment with editing your comment after saving

https://user-images.githubusercontent.com/20773163/112485632-0cc62600-8d73-11eb-8127-5e6d39450890.mp4


### UI accessibility concerns?

As previously noted in the autocomplete work, there are definitely accessibility considerations. This reworked approach has a nice accessibility side effect of only triggering the combobox when the `@mention` has begun. This means when a user types the `@` sign, the screen reader announces that we're now in a combobox. This is similar to the behaviour observed in e.g. GitHub's comments. The previously tested combobox functionality continues to work for screen reader users.

tl;dr I think this is actually a bit better for accessibility than the original implementation.

A screen recording of me testing this out in Safari with VoiceOver:


https://user-images.githubusercontent.com/20773163/112485270-a93bf880-8d72-11eb-8134-141fc0ee2f4a.mp4



## Added tests?

- [X] Yes: Updated the original tests and added a new test for pre-population of comments being edited


## [Forem core team only] How will this change be communicated?

- [X] This change does not need to be communicated, and this is why not: we haven't yet shared the autocomplete functionality in a changelog


## [optional] What gif best describes this PR or how it makes you feel?

![dog rolls around saying work work](https://media2.giphy.com/media/lowAFsBujmjQ0MQ55i/giphy.gif?cid=ecf05e47272hj03hngnw3b99nqxjmcwge0vjp0vednko61xg&rid=giphy.gif)
